File: usrbincrash/usrbincrash.c
#include <stdio.h>
#include <stdlib.h>

long gcd(long x, long y) {
  long t = 0;
  while (y != 0) {
    t = y;
    y = x % y;
    x = t;
  }
  return x;
}

typedef struct { long w; long c; } item;

int main(int argc, char** argv) {
  /* Variable declarations */
  int n = 0;
  int capacity = 50;
  long weight, cost, target, g, i, j, best, c;
  long * table;
  item * items;
  item x;
  FILE *file = fopen(argv[1], "r");
  char* line = malloc(sizeof(char)*255);

  /* Ensure proper input file */
  if ((argc < 1) || (file == NULL)) {
    printf("usage: ./usrbincrash <in>\n");
    free(line);
    exit(1);
  }

  items = malloc(capacity*sizeof(item));
  if (items == NULL) exit(1);
  /* Load the file, ignoring SKU, into struct */
  fscanf(file, "%ld", &target);
  while (fscanf(file, "%s%ld%ld", line, &weight, &cost) == 3) {
    if (n == capacity) {
      capacity *= 2;
      items = realloc(items, capacity*sizeof(item));
      if (items == NULL)
        exit(1);
    }
    x.c = cost;
    x.w = weight;
    items[n] = x;
    n++;
  }
  free(line);
  fclose(file);

  /* Reduce all weights by the GCD */
  g = gcd(target, items[0].w);
  for (i = 1; i < n; i++)
    g = gcd(g, items[i].w);
  if (g > 1) {
    for (i = 0; i < n; i++ ) {
      items[i].w /= g;
    }
    target /= g;
  }

  /* Dynamic Programming Algorithm */
  best = -1;
  table = malloc(sizeof(long) * target);
  for (i = 0; i < target; i++) {
    best = -1;
    for (j = 0; j < n; j++){
      c = items[j].c;
      if (items[j].w <= i)
        c += table[i - items[j].w];
      if ((c < best) || (best == -1))
        best = c;
    } 
    table[i] = best;
  }
  printf("%ld\n", table[target-1]);
  free(items);
  return 0;
}

File: usrbincrash/usrbincrash.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class usrbincrash {

  private static int capacity;
  private static int size;
  private static int[] cost;
  private static int[] weight;

  public static void main(String[] args) throws IOException {
    capacity = 50;
    if (args.length == 0) {
      System.out.println("usage: ./usrbincrash <file>");
      System.exit(0);
    }

    BufferedReader file = new BufferedReader(
      new FileReader(args[0]));

    String temp;
    size = 0;
    int target = Integer.parseInt(file.readLine());
    cost = new int[capacity];
    weight = new int[capacity];

    while((temp=file.readLine()) != null) {
      if (size == capacity) {
        cost = expandArray(cost);
        weight = expandArray(weight);
        capacity *= 2;
      }
      String[] tempa = temp.split("\\s+");
      weight[size] = Integer.parseInt(tempa[1]);
      cost[size] = Integer.parseInt(tempa[2]);
      size++;
    }

    int g = gcd(target,weight[0]);
    for (int i = 1; i < size; i++) {
      g = gcd(g, weight[i]);
    }

    if (g > 1) {
      for (int i = 0; i < size; i++) {
        weight[i] /= g;
      }
      target /= g;
    }

    int best = -1;
    int c;
    int[] table = new int[target];
    for (int i = 0; i < target; i++) {
      best = -1;
      for (int j = 0; j < size; j++) {
        c = cost[j];
        if (weight[j] <= i) {
          c += table[i - weight[j]];
        }
        if ((c < best) || (best == -1)) {
          best = c;
        }
      }
      table[i] = best;
    }
    System.out.println(table[target-1]);
  }

  /*
  *  Euclidean algorithm for GCD
  */
  private static int gcd (int a, int b) {
    int t = 0;
    while (b != 0) {
      t = b;
      b = a % b;
      a = t;
    }
    return a;
  }

  /*
  * expandArray doubles array length if necessary
  */
  private static int[] expandArray(int[] oldArray) {
    int[] newArray = new int[capacity*2];
    for (int i = 0; i < size; i++) {
      newArray[i] = oldArray[i];
    }
    return newArray;
  }
}
